### PR TITLE
Refactor ChatMember methods

### PR DESCRIPTION
**Pull request by WaffleLapkin**
- Add `Deref` impl for `ChatMember` to allow calling kind's methods directly
- Add is_* methods to `ChatMember`
- Make privileges checks easier to use
- Make rights checks easier to use
- Add `ChatMemberKind::is_anonymous` method

### CODE REVIEW

File: src/types/chat_member.rs [hunk 1]
```diff
@@ -127,12 +129,32 @@ pub struct Restricted {
...
+impl Deref for ChatMember {
+    type Target = ChatMemberKind;
+
+    fn deref(&self) -> &Self::Target {
+        &self.kind
+    }
 }
 
-impl ChatMember {
+/// Simple methods to check user status.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:153]:
```suggestion
/// Simple methods for checking a user status.
```

File: src/types/chat_member.rs [hunk 2]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
         }
     }
 
+    /// Returns `true`, if the user's presence in the chat is hidden.
+    ///
+    /// I.e. returns `true` if the user is the creator of the chat or an
+    /// administrator in the chat and has [`can_manage_chat`] privilege.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:250]:
```suggestion
    /// administrator in the chat and has the [`can_manage_chat`] privilege.
```

File: src/types/chat_member.rs [hunk 3]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
             | Self::Kicked(Kicked { until_date, .. }) => Some(*until_date),
         }
     }
+}
 
-    /// Getter for [`Administrator::can_be_edited`] field.
-    pub fn can_be_edited(&self) -> Option<bool> {
-        match &self {
-            Self::Administrator(Administrator { can_be_edited, .. }) => Some(*can_be_edited),
+/// Methods to check admin privileges.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:272]:
```suggestion
/// Methods for checking admin privileges.
```

File: src/types/chat_member.rs [hunk 4]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
         }
     }
 
-    /// Getter for [`Administrator::can_manage_chat`] field.
-    pub fn can_manage_chat(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, if the user can access the chat event log, chat
+    /// statistics, message statistics in channels, see channel members, see
+    /// anonymous administrators in supergroups and ignore slow mode. Implied by
+    /// any other administrator privilege.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:291]: Consider combining the enumeration into a list.

File: src/types/chat_member.rs [hunk 5]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
 
-    /// Getter for [`Administrator::can_change_info`] field.
-    pub fn can_change_info(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can change the chat
+    /// title, photo and other settings.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat
+    /// - is an administrator in the given chat and has [`can_change_info`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:315]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_change_info`]
```

File: src/types/chat_member.rs [hunk 6]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_change_info,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_post_messages`] field.
-    pub fn can_post_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can post in the
```
Comment (Hirrolot) [inline src/types/chat_member.rs:330]:
```suggestion
    /// Returns `true` if the user can post in the
```

File: src/types/chat_member.rs [hunk 7]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
 
-    /// Getter for [`Administrator::can_edit_messages`] field.
-    pub fn can_edit_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can edit messages of
+    /// other users and can pin messages, channels only.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat (even if the chat is not a channel)
+    /// - is an administrator in the given chat and has [`can_edit_messages`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:355]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_edit_messages`]
```

File: src/types/chat_member.rs [hunk 8]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
 
-    /// Getter for [`Administrator::can_delete_messages`] field.
-    pub fn can_delete_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can delete messages
+    /// of other users.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat
+    /// - is an administrator in the given chat and has [`can_delete_messages`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:375]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_delete_messages`]
```

File: src/types/chat_member.rs [hunk 9]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
 
-    /// Getter for [`Administrator::can_manage_voice_chats`] field.
-    pub fn can_manage_voice_chats(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can manage voice chats.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat
+    /// - is an administrator in the given chat and has
+    ///   [`can_manage_voice_chats`] privilege.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:396]:
```suggestion
    /// the [`can_manage_voice_chats`] privilege.
```

File: src/types/chat_member.rs [hunk 10]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
     }
 
-    /// Getter for [`Administrator::can_invite_users`] field.
-    pub fn can_invite_users(&self) -> Option<bool> {
+    /// Returns `true`, the user can can invite new users
+    /// to the chat.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat
+    /// - is an administrator in the given chat and has [`can_invite_users`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:416]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_invite_users`]
```

File: src/types/chat_member.rs [hunk 11]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
 
-    /// Getter for [`Administrator::can_restrict_members`] field.
-    pub fn can_restrict_members(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can restrict,
+    /// ban or unban chat members.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat
+    /// - is an administrator in the given chat and has [`can_restrict_members`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:436]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_restrict_members`]
```

File: src/types/chat_member.rs [hunk 12]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_restrict_members,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_pin_messages`] field.
-    pub fn can_pin_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can pin messages,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:452]:
```
    /// Returns `true` if the user can pin messages,
```

File: src/types/chat_member.rs [hunk 13]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_invite_users,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_restrict_members`] field.
-    pub fn can_restrict_members(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can restrict,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:431]:
```suggestion
    /// Returns `true` if the user can restrict,
```

File: src/types/chat_member.rs [hunk 14]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Left
-            | Self::Kicked(_) => None,
+            }) => *can_manage_voice_chats,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_invite_users`] field.
-    pub fn can_invite_users(&self) -> Option<bool> {
+    /// Returns `true`, the user can can invite new users
```
Comment (Hirrolot) [inline src/types/chat_member.rs:411]:
```suggestion
    /// Returns `true` if the user can can invite new users
```

File: src/types/chat_member.rs [hunk 15]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_delete_messages,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_manage_voice_chats`] field.
-    pub fn can_manage_voice_chats(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can manage voice chats.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:391]:
```suggestion
    /// Returns `true` if the user can manage voice chats.
```

File: src/types/chat_member.rs [hunk 16]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => can_edit_messages.unwrap_or_default(),
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_delete_messages`] field.
-    pub fn can_delete_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can delete messages
```
Comment (Hirrolot) [inline src/types/chat_member.rs:370]:
```suggestion
    /// Returns `true` if the user can delete messages
```

File: src/types/chat_member.rs [hunk 17]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => can_post_messages.unwrap_or_default(),
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_edit_messages`] field.
-    pub fn can_edit_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can edit messages of
```
Comment (Hirrolot) [inline src/types/chat_member.rs:350]:
```suggestion
    /// Returns `true` if the user can edit messages of
```

File: src/types/chat_member.rs [hunk 18]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_manage_chat,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => true,
         }
     }
 
-    /// Getter for [`Administrator::can_change_info`] field.
-    pub fn can_change_info(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can change the chat
```
Comment (Hirrolot) [inline src/types/chat_member.rs:310]:
```suggestion
    /// Returns `true` if the user can change the chat
```

File: src/types/chat_member.rs [hunk 19]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
             | Self::Left
-            | Self::Kicked(_) => None,
+            | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_manage_chat`] field.
-    pub fn can_manage_chat(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, if the user can access the chat event log, chat
```
Comment (Hirrolot) [inline src/types/chat_member.rs:288]:
```suggestion
    /// Returns `true` if the user can access the chat event log, chat
```

The `<X>, if` form is generally acceptable but in this case is less smooth.

File: src/types/chat_member.rs [hunk 20]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
         }
     }
 
+    /// Returns `true`, if the user's presence in the chat is hidden.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:247]:
```suggestion
    /// Returns `true`, if the user's presence in the chat is hidden.
```

File: src/types/chat_member.rs [hunk 21]
```diff
@@ -102,21 +104,21 @@ pub struct Administrator {
 pub struct Restricted {
     /// Date when restrictions will be lifted for
     /// this user, unix time.
-    pub until_date: i32,
+    pub until_date: i64,
 
-    /// Restricted only. `true`, if the user can send text messages,
+    /// `true`, if the user can send text messages,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:109]:
```suggestion
    /// `true`, if the user can send text messages,
```

File: src/types/chat_member.rs [hunk 22]
```diff
@@ -102,21 +104,21 @@ pub struct Administrator {
...
     /// contacts, locations and venues.
     pub can_send_messages: bool,
 
-    /// Restricted only. `true`, if the user is allowed to send audios,
+    /// `true`, if the user is allowed to send audios,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:113]:
```suggestion
    /// `true` if the user is allowed to send audios,
```

File: src/types/chat_member.rs [hunk 23]
```diff
@@ -102,21 +104,21 @@ pub struct Administrator {
...
     /// documents, photos, videos, video notes and voice notes.
     pub can_send_media_messages: bool,
 
-    /// Restricted only. `true`, if the user is allowed to send animations,
+    /// `true`, if the user is allowed to send animations,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:117]:
```suggestion
    /// `true` if the user is allowed to send animations,
````

File: src/types/chat_member.rs [hunk 24]
```diff
@@ -102,21 +104,21 @@ pub struct Administrator {
...
     /// games, stickers and use inline bots.
     pub can_send_other_messages: bool,
 
-    /// Restricted only. `true`, if the user is allowed to add web page
+    /// `true`, if the user is allowed to add web page
```
Comment (Hirrolot) [inline src/types/chat_member.rs:121]:
```suggestion
    /// `true` if the user is allowed to add web page
````

File: src/types/chat_member.rs [hunk 25]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
     }
 
-    /// Getter for [`Administrator::can_pin_messages`] field.
-    pub fn can_pin_messages(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can pin messages,
+    /// supergroups only.
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat (even if the chat is not a supergroups)
```
Comment (Hirrolot) [inline src/types/chat_member.rs:456]:
```suggestion
    /// - is the creator of the chat (even if the chat is not a supergroup)
```

File: src/types/chat_member.rs [hunk 26]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+    /// - is an administrator in the given chat and has [`can_pin_messages`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:457]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_pin_messages`]
```

File: src/types/chat_member.rs [hunk 27]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => can_pin_messages.unwrap_or_default(),
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Administrator::can_promote_members`] field.
-    pub fn can_promote_members(&self) -> Option<bool> {
-        match &self {
+    /// Returns `true`, the user can add new
```
Comment (Hirrolot) [inline src/types/chat_member.rs:472]:
```suggestion
    /// Returns `true` if the user can add new
```

File: src/types/chat_member.rs [hunk 28]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+    /// administrators with a subset of his own privileges or demote
+    /// administrators that he has promoted, directly or indirectly
+    /// (promoted by administrators that were appointed by the
+    /// user).
+    ///
+    /// I.e. returns `true` if the user
+    /// - is the creator of the chat (even if the chat is not a channel)
+    /// - is an administrator in the given chat and has [`can_promote_members`]
```
Comment (Hirrolot) [inline src/types/chat_member.rs:480]:
```suggestion
    /// - is an administrator in the given chat and has the [`can_promote_members`]
```

File: src/types/chat_member.rs [hunk 29]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_promote_members,
+            Self::Member | Self::Restricted(_) | Self::Left | Self::Kicked(_) => false,
         }
     }
+}
 
-    /// Getter for [`Restricted::can_send_messages`] field.
-    pub fn can_send_messages(&self) -> Option<bool> {
+/// Methods to check member rights.
```
Comment (Hirrolot) [inline src/types/chat_member.rs:497]:
```suggestion
    /// Methods for checking member rights.
```

File: src/types/chat_member.rs [hunk 30]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+impl ChatMemberKind {
+    /// Returns `true`, if the user can send text messages,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:499]:
```suggestion
    /// Returns `true` if the user can send text messages,
```

File: src/types/chat_member.rs [hunk 31]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+    /// contacts, locations and venues.
+    ///
+    /// I.e. returns **`false`** if the user
+    /// - has left or has been kicked from the chat
+    /// - is restricted and doesn't have [`can_send_messages`] right
```
Comment (Hirrolot) [inline src/types/chat_member.rs:504]:
```suggestion
    /// - is restricted and doesn't have the [`can_send_messages`] right
```

File: src/types/chat_member.rs [hunk 32]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_send_messages,
+            Self::Creator(_) | Self::Administrator(_) | Self::Member => true,
+            Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Restricted::can_send_media_messages`] field.
-    pub fn can_send_media_messages(&self) -> Option<bool> {
+    /// Returns `true`, if the user is allowed to send audios,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:518]:
```suggestion
    /// Returns `true` if the user is allowed to send audios,
```

File: src/types/chat_member.rs [hunk 33]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+    /// documents, photos, videos, video notes and voice notes.
+    ///
+    /// I.e. returns **`false`** if the user
+    /// - has left or has been kicked from the chat
+    /// - is restricted and doesn't have [`can_send_media_messages`] right
```
Comment (Hirrolot) [inline src/types/chat_member.rs:523]:
```suggestion
    /// - is restricted and doesn't have the [`can_send_media_messages`] right
```

File: src/types/chat_member.rs [hunk 34]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_send_media_messages,
+            Self::Creator(_) | Self::Administrator(_) | Self::Member => true,
+            Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Restricted::can_send_other_messages`] field.
-    pub fn can_send_other_messages(&self) -> Option<bool> {
+    /// Returns `true`, if the user is allowed to send animations,
```
Comment (Hirrolot) [inline src/types/chat_member.rs:538]:
```suggestion
    /// Returns `true` if the user is allowed to send animations,
```

File: src/types/chat_member.rs [hunk 35]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+    /// games, stickers and use inline bots.
+    ///
+    /// I.e. returns **`false`** if the user
+    /// - has left or has been kicked from the chat
+    /// - is restricted and doesn't have [`can_send_media_messages`] right
```
Comment (Hirrolot) [inline src/types/chat_member.rs:543]:
```suggestion
    /// - is restricted and doesn't have the [`can_send_media_messages`] right
```

File: src/types/chat_member.rs [hunk 36]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
-            | Self::Kicked(_) => None,
+            }) => *can_send_other_messages,
+            Self::Creator(_) | Self::Administrator(_) | Self::Member => true,
+            Self::Left | Self::Kicked(_) => false,
         }
     }
 
-    /// Getter for [`Restricted::can_add_web_page_previews`] field.
-    pub fn can_add_web_page_previews(&self) -> Option<bool> {
+    /// Returns `true`, if the user is allowed to add web page
```
Comment (Hirrolot) [inline src/types/chat_member.rs:558]:
```suggestion
    /// Returns `true` if the user is allowed to add web page
```

File: src/types/chat_member.rs [hunk 37]
```diff
@@ -154,227 +244,334 @@ impl ChatMemberKind {
...
+    /// previews to their messages.
+    ///
+    /// I.e. returns **`false`** if the user
+    /// - has left or has been kicked from the chat
+    /// - is restricted and doesn't have [`can_send_media_messages`] right
```
Comment (Hirrolot) [inline src/types/chat_member.rs:563]:
```suggestion
    /// - is restricted and doesn't have the [`can_send_media_messages`] right
```

**Review by Hirrolot — CHANGES_REQUESTED** [review]

File: src/types/chat_member.rs (hunk 4 above)
Comment (WaffleLapkin) [reply src/types/chat_member.rs:291]: I don't think it's a good idea to use lists in the first sentence.

**Review by Hirrolot — APPROVED** [review]

**Pull request merged**